File: platform/src/audit_platform/utils/atomic_write.py
"""Atomic, backed-up JSON writer for pipeline data files.

Python twin of ``template/scripts/lib/atomic-write.js`` (see lines 35-72 there).
A crash, signal, or concurrent write between flush and rename cannot leave the
target in a partial/invalid state. The prior version is preserved at
``${target}.bak`` for recovery.

Flow:
    1. If target exists, copy it to ``${target}.bak`` (overwrites any prior .bak).
    2. Write JSON to ``${target}.tmp-<pid>-<ms>``, flush, and ``os.fsync`` the fd.
    3. ``os.replace(tmp, target)`` — atomic POSIX rename within a filesystem.

On any exception between steps 1 and 3: the target bytes are untouched, the
``.bak`` is preserved, and the ``.tmp-*`` file is best-effort removed.
"""

from __future__ import annotations

import json
import os
import shutil
import time
from pathlib import Path
from typing import Any


def write_json_atomic(target: Path | str, data: Any, *, indent: int = 2) -> None:
    """Write ``data`` as JSON to ``target`` atomically with a ``.bak`` of the prior file.

    Args:
        target: Path to the final file (str or Path).
        data: Any JSON-serializable value.
        indent: ``json.dumps`` indent (default 2). Pass ``None`` for compact output.
    """
    target_path = Path(target).resolve()
    bak_path = target_path.with_name(target_path.name + ".bak")
    tmp_path = target_path.with_name(
        f"{target_path.name}.tmp-{os.getpid()}-{int(time.time() * 1000)}"
    )

    # 1. Back up prior version if present.
    if target_path.exists():
        shutil.copy2(target_path, bak_path)

    # 2. Write tmp + fsync so the bytes are durable before the rename.
    try:
        with open(tmp_path, "w", encoding="utf-8") as fh:
            json.dump(data, fh, indent=indent, default=str)
            fh.flush()
            os.fsync(fh.fileno())
    except Exception:
        tmp_path.unlink(missing_ok=True)
        raise

    # 3. Atomic rename. On failure, drop the tmp; .bak + original target are intact.
    try:
        os.replace(tmp_path, target_path)
    except Exception:
        tmp_path.unlink(missing_ok=True)
        raise

File: clients/laura-willis/scripts/lib/atomic-write.js
'use strict';

/**
 * atomic-write.js — Atomic, backed-up JSON writer for pipeline data files.
 *
 * Guarantees: a crash, signal, or concurrent write cannot leave the target
 * in a partial/invalid state. The prior version is preserved in a fixed
 * `${target}.bak` for recovery.
 *
 * Flow:
 *   1. If target exists, copy it to `${target}.bak` (overwrites any prior .bak).
 *   2. Write JSON to `${target}.tmp-<pid>-<ts>` and fsync the fd.
 *   3. fs.renameSync(tmp, target) — POSIX atomic within a filesystem.
 *
 * On any throw between (1) and (3): target bytes are untouched, .bak is
 * preserved, and the .tmp file is best-effort unlinked.
 *
 * Usage:
 *   const { writeJsonAtomic } = require('./lib/atomic-write');
 *   writeJsonAtomic('seo/audit-data.json', data, { indent: 2 });
 *   writeJsonAtomic('seo/audit-data.json', data, { indent: 2, trailingNewline: true });
 */

const fs = require('fs');
const path = require('path');

/**
 * Write `value` as JSON to `target` atomically, with a `.bak` of the prior
 * file preserved on disk.
 *
 * @param {string} target - Absolute or relative path to the final file.
 * @param {*} value - Any JSON-serializable value.
 * @param {{ indent?: number, trailingNewline?: boolean }} [options]
 */
function writeJsonAtomic(target, value, options) {
  const opts = options || {};
  const indent = opts.indent == null ? 2 : opts.indent;
  const trailingNewline = opts.trailingNewline === true;

  const targetAbs = path.resolve(target);
  const bakPath = targetAbs + '.bak';
  const tmpPath = targetAbs + '.tmp-' + process.pid + '-' + Date.now();

  let serialized = JSON.stringify(value, null, indent);
  if (trailingNewline) serialized += '\n';

  // 1. Back up prior version if present.
  if (fs.existsSync(targetAbs)) {
    fs.copyFileSync(targetAbs, bakPath);
  }

  // 2. Write tmp + fsync.
  let fd;
  try {
    fd = fs.openSync(tmpPath, 'w');
    fs.writeSync(fd, serialized);
    fs.fsyncSync(fd);
  } catch (err) {
    try { if (fd != null) fs.closeSync(fd); } catch (_) { /* noop */ }
    try { fs.unlinkSync(tmpPath); } catch (_) { /* noop */ }
    throw err;
  }
  fs.closeSync(fd);

  // 3. Atomic rename.
  try {
    fs.renameSync(tmpPath, targetAbs);
  } catch (err) {
    try { fs.unlinkSync(tmpPath); } catch (_) { /* noop */ }
    throw err;
  }
}

module.exports = { writeJsonAtomic };
